src/backend/distributed/utils/node_metadata.c

```diff
@@ -454,6 +456,103 @@ ActivateNode(char *nodeName, int nodePort)
 
 
 /*
+ * master_update_node moves the requested node to a different nodename and nodeport. It
+ * locks to ensure no queries are running concurrently; and is intended for customers who
+ * are running their own failover solution.
+ */
+Datum
+master_update_node(PG_FUNCTION_ARGS)
+{
+	int32 nodeId = PG_GETARG_INT32(0);
+
+	text *newNodeName = PG_GETARG_TEXT_P(1);
+	int32 newNodePort = PG_GETARG_INT32(2);
+
+	char *newNodeNameString = text_to_cstring(newNodeName);
+
+	CheckCitusVersion(ERROR);
+
+	/*
+	 * This lock has two purposes:
+	 * - Ensure buggy code in Citus doesn't cause failures when the nodename/nodeport of
+	 *   a node changes mid-query
+	 * - Provide fencing during failover, after this function returns all connections
+	 *   will use the new node location.
+	 *
+	 * Drawback:
+	 * - This function blocks until all previous queries have finished. This means that
+	 *   long-running queries will prevent failover.
+	 */
+	LockRelationOid(DistNodeRelationId(), ExclusiveLock);
```

Comment:
I think this should be `AccessExclusiveLock` to actually provide fencing. Even writes only take an `AccessShareLock` on `pg_dist_node`

Reply:
Yes, you're right